src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasures.ts

```diff
@@ -0,0 +1,123 @@
+  id: any,
+): string[] => {
+  const { structure } = legacyCalculatedMeasureFolder.children ?? {};
+  const ids = Object.keys(
+    legacyCalculatedMeasureFolder.children?.content.children ?? {},
+  );
+  const contentRecords = structure ? findContentRecords(structure, ids) : {};
+
+  const calculatedMeasureName = JSON.parse(
+    getMetaData(contentRecords[id].node, id),
+  ).name;
+
+  return calculatedMeasureName;
+};
+
+/**
+ * Transforms the serialized definitions of all calculated measures created with ActiveUI 5.0 and used in a saved dashboard or saved widget, into ones that are natively supported by ActivePivot.
+ * Removes any query-scoped calculated measure definitions from saved dashboards and saved widgets.
+ * Mutates `contentServer`.
+ */
+export function migrateCalculatedMeasures(
+  contentServer: ContentRecord,
+  dataModel: DataModel,
+): any {
+  const legacyCalculatedMeasuresFolder =
+    contentServer.children?.ui.children?.calculated_measures;
+  const cmFolder: ContentRecord | undefined =
+    contentServer.children?.pivot.children?.entitlements.children?.cm;
+
+  const legacyCalculatedMeasureRecords: {
+    [measureName: string]: ContentRecord;
+  } = legacyCalculatedMeasuresFolder?.children
+    ? _mapKeys(
+        legacyCalculatedMeasuresFolder.children.content.children,
+        (value, key) =>
+          getCalculatedMeasureName(legacyCalculatedMeasuresFolder, key),
+      )
+    : {};
+
+  const namesOfCalculatedMeasuresToMigrate = Object.keys(
+    legacyCalculatedMeasureRecords,
+  );
+
+  const {
+    migratedWidgetsRecord,
+    measureToCubeMapping: measureToCubeMappingInWidgets,
+  } = migrateCalculatedMeasuresInWidgets(
+    contentServer.children?.ui.children?.widgets!,
+    dataModel,
+    namesOfCalculatedMeasuresToMigrate,
+  );
+
+  const {
+    migratedDashboards,
+    measureToCubeMapping: measureToCubeMappingInDashboards,
+  } = migrateCalculatedMeasuresInDashboards(
+    contentServer.children?.ui.children?.dashboards!,
+    dataModel,
+    namesOfCalculatedMeasuresToMigrate,
+  );
+
+  const measureToCubeMapping = _merge(
+    measureToCubeMappingInWidgets,
+    measureToCubeMappingInDashboards,
+  );
+
+  Object.entries(legacyCalculatedMeasureRecords).forEach(
+    ([measureName, record]) => {
+      const cubeNames = measureToCubeMapping[measureName];
+      if (!cubeNames) {
+        // TODO add a warning in the report that the calculated measure was not migrated since it's not used anywhere.
+        return;
+      }
+
+      const migratedContent = migrateCalculatedMeasureContent(
+        JSON.parse(record.entry.content),
+        measureName,
+      );
+      record.entry.content = JSON.stringify(migratedContent);
+
+      cubeNames.forEach((cubeName) => {
+        if (cmFolder && cmFolder.children) {
+          // If `cubeName` property does not already exist, create it.
+          if (!cmFolder.children[cubeName]) {
+            cmFolder.children[cubeName] = {
```

Comment:
Will the `cubeName` already exist in the `pivot/entitlements/cm` folder or will there be instances where we need to create it ? I can delete this if it is not necessary.